### PR TITLE
Use conditions but not repo ids as query condition

### LINKED ISSUE #15412: Open Pull-Requests view with > 500 organizations need more than 1 minute
- Gitea version (or commit ref): 1.14.0
- Git version:
- Operating system: 20.04.2
  
  
  
- Database (use `[x]`):
  - [x] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
- Log gist:

## Description

This with 1 minute is on my internal mini-server configuration.

I try to analyze what are you doing after I open the Pull-Requests link.

image

What I see.

First this PostgreSQL query:

`explain analyze SELECT "repository"."id" FROM "repository" INNER JOIN "team_user" ON repository.owner_id = team_user.org_id INNER JOIN "team_repo" ON ('true' != 'false' and repository.is_private != 't') OR (team_user.team_id = team_repo.team_id AND repository.id = team_repo.repo_id) WHERE (team_user.uid = '1') GROUP BY repository.id`

image

Which needs ~7 seconds on my NUC (https://ark.intel.com/content/www/us/en/ark/products/126137/intel-nuc-kit-nuc7pjyh.html -> I'm running here Gitea 1.14.0, PostgreSQL 13, Nginx, InfluxDB, Grafana), And after the first query I see a lot small PostgreSQL queries. I think one for every organization.

And when I choose other organization from the long list

image

I see > 500 http requests to the Gitea server

image

Looks like the process is "complicated" and not scalable for a lot of organizations and repos. Cause I have no pull requests on my Gitea instance

image

I think you should return back only all organizations which has pull requests. In my case: no one. Just the info: you dind't have any pull requests.

But this "problem" has only user with a lot of organizations and repos.

Am I right?

### PR DESCRIPTION
**Pull request by lunny**
Fix #15412, in my test, over `600` repositories in `/issues` with over `6000` issues will spend `370ms` in my macBook Pro 2020 .

### CODE REVIEW

**Comment by zeripath** [comment]
Ci fail appears related

**Comment by lunny** [comment]
> Ci fail appears related

Yes, I'm looking into that. A question is `if I created an issue or pull in a public repository I have no write permission, should the repo be listed in my issues or pulls`. Or any public repo which you have some comments/reviews on those issues or pulls.

**Comment by wxiaoguang** [comment]
It's more clear than before 👍.

I am a little curious, why these unit test checks for `IssueStats` are changed?

File: routers/web/user/home.go [hunk 1]
```diff
@@ -797,33 +692,27 @@ func issueIDsFromSearch(ctxUser *user_model.User, keyword string, opts *models.I
...
-			if repo_model.IsErrRepoNotExist(err) {
+		repoIDs = append(repoIDs, id)
+		if len(repoIDs) == 500 {
+			if err := repo_model.FindReposMapByIDs(repoIDs, totalRes); err != nil {
 				return nil, err
-			} else if err != nil {
-				return nil, fmt.Errorf("GetRepositoryByID: [%d]%v", id, err)
 			}
-			repoByID[id] = repo
+			repoIDs = make([]int64, 0, 500)
```
Comment (zeripath) [inline routers/web/user/home.go]: why not:

```suggestion
			repoIDs = repoIDs[:0]
```

**Review by zeripath — APPROVED** [review]
small nit otherwise LGTM

File: models/issue.go [hunk 2]
```diff
@@ -1329,6 +1332,44 @@ func (opts *IssuesOptions) setupSession(sess *xorm.Session) {
...
+			cond = cond.And(
+				builder.Or(
+					userOrgUnitRepoCond(repoIDstr, userID, org.ID, unitType), // team member repos
+					userOrgPublicUnitRepoCond(userID, org.ID),                // user org public non-member repos, TODO: check repo has issues
+				),
+			)
+		}
+	} else {
+		cond = cond.And(
+			builder.Or(
```
Comment (zeripath) [inline models/issue.go:1371]: At some point we need to move to unions for some of these.

Comment (lunny) [reply models/issue.go:1371]: But not this PR. :)

**Comment by lunny** [comment]
> It's more clear than before 👍.
> 
> I am a little curious, why these unit test checks for `IssueStats` are changed?

I guess the previous implementation is not right. :) But I cannot point out what's wrong with it.
I tested the new implementation locally and found it works well.

File: models/repo_list.go [hunk 3]
```diff
@@ -145,6 +146,180 @@ type SearchRepoOptions struct {
...
+	return builder.In(idStr, builder.Select("repo_id").
+		From("`access`").
+		Where(builder.And(
+			builder.Eq{"user_id": userID},
+			builder.Gt{"mode": int(perm.AccessModeNone)},
+		)),
+	)
+}
+
+func userOrgTeamRepoCond(idStr string, userID int64) builder.Cond {
```
Comment (noerw) [inline models/repo_list.go:269]: Just a drive-by comment: would you mind to add a comment on each function what these conditions/builders do? This will document intended behaviour and help with debugging
eg. `// selects repos that the given user has access to through team membership` (if I understand that query correctly)

Comment (lunny) [reply models/repo_list.go:269]: Done.

File: models/issue.go [hunk 4]
```diff
@@ -1663,28 +1705,32 @@ func GetUserIssueStats(opts UserIssueStatsOptions) (*IssueStats, error) {
 		cond = cond.And(builder.In("issue.id", opts.IssueIDs))
 	}
 
+	if opts.UserID > 0 {
+		cond = cond.And(issuePullAccessibleRepoCond("issue.repo_id", opts.UserID, opts.Org, opts.Team, opts.IsPull))
+	}
+
 	sess := func(cond builder.Cond) *xorm.Session {
 		s := db.GetEngine(db.DefaultContext).Where(cond)
+		s.Join("INNER", "repository", "issue.repo_id = repository.id")
```
Comment (wxiaoguang) [inline models/issue.go]: Why should this be moved out of `if opts.IsArchived != util.OptionalBoolNone`?

File: models/repo_list.go [hunk 5]
```diff
@@ -145,6 +146,188 @@ type SearchRepoOptions struct {
...
+			}.And(
+				builder.In(
+					"team_id", builder.Select("team_id").From("team_user").Where(
+						builder.Eq{
+							"uid": userID,
+						},
+					),
+				)).And(
+				builder.In(
+					"team_id", builder.Select("team_id").From("team_unit").Where(
```
Comment (wxiaoguang) [inline models/repo_list.go:245]: Not sure whether it is better to use `team_user JOIN team_unit` (this is also fine to me)

File: models/repo_list.go [hunk 6]
```diff
@@ -145,6 +146,188 @@ type SearchRepoOptions struct {
...
+func userOrgPublicRepoCondPrivate(userID int64) builder.Cond {
+	return builder.And(
+		builder.Eq{"`repository`.is_private": false},
+		builder.In("`repository`.owner_id",
+			builder.Select("`org_user`.org_id").
+				From("org_user").
+				Join("INNER", "`user`", "`user`.id = `org_user`.org_id").
+				Where(builder.Eq{
+					"`org_user`.uid":    userID,
+					"`user`.type":       user_model.UserTypeOrganization,
```
Comment (wxiaoguang) [inline models/repo_list.go]: `type` may be a keyword? Should it be quoted?

File: models/repo_list.go (hunk 5 above)
Comment (lunny) [reply models/repo_list.go:245]: Subquery is slightly faster than Join. And this also makes the code look clear.

File: models/repo_list.go (hunk 6 above)
Comment (lunny) [reply models/repo_list.go]: Good Catch.

**Review by wxiaoguang — APPROVED** [review]
Generally LGTM. Quite complex, maybe we then should try it in production to have further tests.
